prepare_stmt.go

```diff
@@ -48,22 +46,32 @@ func (db *PreparedStmtDB) Close() {
 	db.Mux.Lock()
 	defer db.Mux.Unlock()
```

Comment:
Dose write lock necessary? It seems that only `db.Stmts = nil` writes?

Reply:
if we don't use write lock, then there would be data race condition, it can be easily proved by running following codes(run by `go run -race main.go`):
```golang
package main

import (
        //"fmt"
        "sync"
)

func main() {
        var m map[string]int
        var lock sync.RWMutex
        var wg sync.WaitGroup
        loopCount := 100000

        wg.Add(2)
        go func() {
                defer wg.Done()
                var val int
                for i := 0; i < loopCount; i++ {
                        lock.RLock()
                        val += m["not exist"]
                        lock.RUnlock()
                }
        }()
        go func() {
                defer wg.Done()
                for i := 0; i < loopCount; i++ {
                        lock.RLock()
                        m = make(map[string]int, 0)
                        lock.RUnlock()
                }
        }()

        wg.Wait()
}
```

Reply:
> if we don't use write lock, then there would be data race condition, it can be easily proved by running following codes(run by `go run -race main.go`):
> 
> ```go
> package main
> 
> import (
>         //"fmt"
>         "sync"
> )
> 
> func main() {
>         var m map[string]int
>         var lock sync.RWMutex
>         var wg sync.WaitGroup
>         loopCount := 100000
> 
>         wg.Add(2)
>         go func() {
>                 defer wg.Done()
>                 var val int
>                 for i := 0; i < loopCount; i++ {
>                         lock.RLock()
>                         val += m["not exist"]
>                         lock.RUnlock()
>                 }
>         }()
>         go func() {
>                 defer wg.Done()
>                 for i := 0; i < loopCount; i++ {
>                         lock.RLock()
>                         m = make(map[string]int, 0)
>                         lock.RUnlock()
>                 }
>         }()
> 
>         wg.Wait()
> }
> ```

I mean, can we use read lock in `for loop` and write lock only when assigning value.

Reply:
No, I don't think we should, chances that some other goroutines could get a closed stmt and ran into the time gap between RUnlock and Lock (if we change codes to the following).
```golang
func (db *PreparedStmtDB) Close() {
	db.Mux.RLock()
	for _, stmt := range db.Stmts {
		go func(s *Stmt) {
			// make sure the stmt must finish preparation first
			<-s.prepared
			if s.Stmt != nil {
				_ = s.Close()
			}
		}(stmt)
	}
        db.Mux.RUnlock()

        db.Mux.Lock()
        defer db.Mux.Unlock()
	// setting db.Stmts to nil to avoid further using
	db.Stmts = nil
}
```
There will be some concurrency problems:
1）the prepare function use RLock too , some other goroutines can run into it while we are closing old Stmt, and they will get a closed/closing/valid Stmt(depends on the range progess)
![image](https://github.com/user-attachments/assets/6d8d3f37-4722-457f-be31-754798c5b4ee)

2）there is a time gap between `db.Mux.RUnlock()` and `db.Mux.Lock()`, and chances that prepare function can get the write lock between this time gap(for example, if there are a goroutine waiting at Line 84), and they could get the WriteLock ahead of the Close function, then trying to insert a new Stmt, which result in Stmt leak:
![image](https://github.com/user-attachments/assets/68f177a2-2e42-4e65-aa46-74d86dabba64)